rpcserver.go

```diff
@@ -6086,65 +6129,76 @@ func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin
 	conn.SetReadDeadline(timeZeroVal)
 
 	// Attempt to parse the raw body into a JSON-RPC request.
-	var responseID interface{}
 	var jsonErr error
-	var result interface{}
-	var request dcrjson.Request
+	var msg []byte
+	var request interface{}
 	if err := json.Unmarshal(body, &request); err != nil {
 		jsonErr = &dcrjson.RPCError{
 			Code: dcrjson.ErrRPCParse.Code,
 			Message: fmt.Sprintf("Failed to parse request: %v",
 				err),
 		}
-	}
-	if jsonErr == nil {
-		// Requests with no ID (notifications) must not have a response
-		// per the JSON-RPC spec.
-		if request.ID == nil {
-			return
-		}
 
-		// The parse was at least successful enough to have an ID so
-		// set it for the response.
-		responseID = request.ID
+		msg, err = createMarshalledReply("", "", jsonErr)
+		if err != nil {
+			rpcsLog.Errorf("Failed to marshal reply: %v", err)
+		}
+	}
 
+	if jsonErr == nil {
 		// Setup a close notifier.  Since the connection is hijacked,
 		// the CloseNotifer on the ResponseWriter is not available.
 		closeChan := make(chan struct{}, 1)
 		go func() {
-			_, err := conn.Read(make([]byte, 1))
+			_, err = conn.Read(make([]byte, 1))
 			if err != nil {
 				close(closeChan)
 			}
 		}()
 
-		// Check if the user is limited and set error if method
-		// unauthorized
-		if !isAdmin {
-			if _, ok := rpcLimited[request.Method]; !ok {
-				jsonErr = rpcInvalidError("limited user not " +
-					"authorized for this method")
+		// determine if the request is single or batched
+		switch request.(type) {
+		case map[string]interface{}:
+			request := request.(map[string]interface{})
+			req, err := dcrjson.NewRequest(request["id"].(interface{}),
+				request["method"].(string), request["params"].([]interface{}))
+			if err != nil {
+				rpcsLog.Errorf("Failed to parse request: %v", err)
+				return
 			}
-		}
+			msg = s.processRequest(req, isAdmin, closeChan)
+		case []interface{}:
+			// process the request as a batched request
+			reqs := request.([]interface{})
+			var batchedReply bytes.Buffer
+			batchedReply.WriteByte('[')
+			for idx, entry := range reqs {
+				entry := entry.(map[string]interface{})
+				req, err := dcrjson.NewRequest(entry["id"].(interface{}),
```

Comment:
This is an example of what I meant by naked type casts that can crash.

Notice what would happen if you fed this code the requests like:

`[{"jsonrpc":1.0", "bogus":"foo"}]` or `[{"jsonrpc":1.0", "id":"1", params:"foo"}]`